File: python/239-Sliding-Window-Maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        output = []
        q = collections.deque() # index
        l = r = 0
        # O(n) O(n)
        while l < len(nums):
            # pop smaller values from q
            while q and nums[q[-1]] < nums[r]:
                q.pop()
            q.append(r)

            # remove left val from window
            if l > q[0]:
                q.popleft()

            if (r + 1) >= k:
                output.append(nums[q[0]])
                l += 1
            r += 1

        return output

File: csharp/239-Sliding-Window-Maximum.cs
public class Solution
{
    public int[] MaxSlidingWindow(int[] nums, int k)
    {

        var output = new List<int>();
        LinkedList<int> queue = new();
        int left = 0, right = 0;

        while (right < nums.Length)
        {
            // pop smaller values from queue
            while (queue.Count > 0 && nums[queue.Last.Value] < nums[right])
                queue.RemoveLast();
            queue.AddLast(right);

            // remove left val from the window
            if (left > queue.First.Value)
                queue.RemoveFirst();

            if (right + 1 >= k)
            {
                output.Add(nums[queue.First.Value]);
                left++;
            }

            right++;
        }

        return output.ToArray();
    }
}